python/src/iceberg/table/metadata.py

```diff
@@ -0,0 +1,259 @@
+# Licensed to the Apache Software Foundation (ASF) under one
+# or more contributor license agreements.  See the NOTICE file
+# distributed with this work for additional information
+# regarding copyright ownership.  The ASF licenses this file
+# to you under the Apache License, Version 2.0 (the
+# "License"); you may not use this file except in compliance
+# with the License.  You may obtain a copy of the License at
+#
+#   http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing,
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+from enum import Enum
+from typing import (
+    Any,
+    Dict,
+    List,
+    Literal,
+    Optional,
+    Union,
+)
+from uuid import UUID
+
+from pydantic import Field, root_validator
+
+from iceberg.schema import Schema
+from iceberg.utils.iceberg_base_model import IcebergBaseModel
+
+_INITIAL_SEQUENCE_NUMBER = 0
+INITIAL_SPEC_ID = 0
+DEFAULT_SCHEMA_ID = 0
+
+
+class SnapshotRefType(str, Enum):
+    branch = "branch"
+    tag = "tag"
+
+
+class SnapshotRef(IcebergBaseModel):
+    snapshot_id: int = Field(alias="snapshot-id")
+    snapshot_ref_type: SnapshotRefType = Field(alias="type")
+    min_snapshots_to_keep: int = Field(alias="min-snapshots-to-keep")
+    max_snapshot_age_ms: int = Field(alias="max-snapshot-age-ms")
+    max_ref_age_ms: int = Field(alias="max-ref-age-ms")
+
+
+class TableMetadataCommonFields(IcebergBaseModel):
```

Comment:
Since refs are added as optional in V2:
![image](https://user-images.githubusercontent.com/1134248/175399608-383f2ff4-3b36-481d-b317-fbb7a2bf177f.png)

I've added them as a property of the V2. Since they are optional, I set the `default_factory` to `dict`:
```
class TableMetadataV2(TableMetadataCommonFields, IcebergBaseModel):
...

    refs: Dict[str, SnapshotRef] = Field(default_factory=dict)
    """A map of snapshot references.
    The map keys are the unique snapshot reference names in the table,
    and the map values are snapshot reference objects.
    There is always a main branch reference pointing to the
    current-snapshot-id even if the refs map is null."""
```
If it isn't supplied, an empty `dict` will be set.

Reply:
Got it, that makes sense. I think we still want to add `main` if `current-snapshot-id` is set, so we probably do want to use a validator.

Reply:
Yes, just added that from the logic in:  https://github.com/apache/iceberg/blob/master/core/src/main/java/org/apache/iceberg/TableMetadataParser.java#L309

Came up with:

```python
    @root_validator(skip_on_failure=True)
    def construct_refs(cls, data: Dict[str, Any]):
        # This is going to be much nicer as soon as sort-order is an actual pydantic object
        if not data.get("refs"):
            if current_snapshot_id := data.get("current_snapshot_id"):
                if current_snapshot_id != -1:
                    data["refs"] = {
                        MAIN_BRANCH: SnapshotRef(snapshot_id=current_snapshot_id, snapshot_ref_type=SnapshotRefType.branch)
                    }
        return data
```

Reply:
Looks good, except for the comment about sort order.